stan/math/prim/prob/wiener4_lcdf.hpp

```diff
@@ -0,0 +1,762 @@
+#ifndef STAN_MATH_PRIM_PROB_WIENER4_LCDF_HPP
+#define STAN_MATH_PRIM_PROB_WIENER4_LCDF_HPP
+
+#include <stan/math/prim/prob/wiener5_lpdf.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Helper function. Calculate \f$\mathbb{e}^{x}\ if \f$x\ is not too large.
+ *
+ * @param x A scalar
+ * @return \f$\mathbb{e}^{x}\
+ */
+template <typename T_x>
+inline auto rexp(T_x&& x) noexcept {
+  return (x <= 700) ? exp(x) : exp(700);
+}
+
+/**
+ * Helper function. Log of Mill's ratio for the normal distribution
+ *
+ * @param x A scalar
+ * @return Log of Mill's ratio
+ */
+template <typename T_x>
+inline auto logMill(T_x&& x) noexcept {
+  if (x > 1.0e5) {
+    return -log(x);
+  }
+  const auto m = std_normal_lcdf(-x) - std_normal_log(x);
+  return m;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale for distribution
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_distribution(const T_a& a, const T_v& v,
+                                         const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
+  if (fabs(v) == 0.0) {
+    return ret_t(log1p(-w));
+  }
+  auto minus_two_va_one_minus_w = (-2.0 * v * a * (1.0 - w));
+  ret_t prob;
+  if (minus_two_va_one_minus_w < 0) {
+    const auto exp_arg = exp(minus_two_va_one_minus_w);
+    if (exp_arg >= nearly_one) {
+      return ret_t(log1p(-w));
+    }
+    auto two_vaw = 2 * v * a * w;
+    if (two_vaw > minus_two_va_one_minus_w) {
+      prob = log1p(-exp_arg) - log_diff_exp(two_vaw, minus_two_va_one_minus_w);
+    } else if (two_vaw < minus_two_va_one_minus_w) {
+      prob = log1p(-exp_arg) - log_diff_exp(minus_two_va_one_minus_w, two_vaw);
+    } else {
+      prob = log1p(-exp_arg) - NEGATIVE_INFTY;
+    }
+  } else {
+    const auto exp_arg = exp(-minus_two_va_one_minus_w);
+    if (exp_arg >= nearly_one)
+      return ret_t(log1p(-w));
+    prob = log1p(-exp_arg) - log1p(-exp(2 * v * a));
+  }
+  return prob;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale for grad_a and grad_v
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_GradAV(const T_a& a, const T_v& v,
+                                   const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
+  if (fabs(v) == 0.0) {
+    return ret_t(-w);
+  }
+  nearly_one = ret_t(1.0 - 1.1 * 1.0e-5);
```

Comment:
Make a test case that shows where this would fail if `nearly_one` was removed

Reply:
> test case that shows where this would fail if `nearly_one` was removed

When I comment the `nearly_one` lines out, then the following error occurs in the test
`./runTests.py -j4 test/unit/math/prim/prob/wiener_full_lcdf_test.cpp`

![image](https://github.com/user-attachments/assets/68e8b02f-0932-41db-9373-c3a2a22ee2e9)


So this `nearly_one` seems to be necessary, what do you think?